demux/utils/novaseq_samplesheet.py

```diff
@@ -0,0 +1,181 @@
+        self.dummy_indexes = dummy_indexes
+        self.runparameters = NovaseqRunParameters(self.flowcell, runs_dir)
+
+    @property
+    def header(self):
+        """ Create the sample sheet header """
+        return list(Samplesheet.header_map.values())
+
+    @staticmethod
+    def get_project(project):
+        """ Only keeps the first part of the project name"""
+        return project.split(" ")[0]
+
+    @staticmethod
+    def reverse_complement(dna):
+        """ Generates the reverse complement """
+        complement = {"A": "T", "C": "G", "G": "C", "T": "A"}
+        return "".join([complement[base] for base in dna[::-1]])
+
+    def add_dummy_indexes(self):
+        """ Add all dummy indexes to raw sample sheet """
+        with open(f"{self.dummy_indexes}") as csv_file:
+            dummy_samples_csv = csv.reader(csv_file, delimiter=",")
+            dummy_samples = [row for row in dummy_samples_csv]
+            added_dummy_samples = []
+            lanes = {sample["lane"] for sample in self.raw_samplesheet}
+
+            for lane in lanes:
+                sample_indexes = [
+                    sample["index"]
+                    for sample in self.raw_samplesheet
+                    if sample["lane"] == lane
+                ]
+                for name, dummy_index in dummy_samples:
+                    if not (
+                        any(
+                            sample_index.startswith(dummy_index)
+                            for sample_index in sample_indexes
+                        )
+                    ):
+                        add_dummy_sample = {
+                            "control": "N",
+                            "description": "",
+                            "fcid": self.flowcell,
+                            "index": dummy_index,
+                            "index2": "",
+                            "lane": lane,
+                            "operator": "script",
+                            "project": "indexcheck",
+                            "recipe": "R1",
+                            "sample_id": name.replace(" ", "-")
+                            .replace("(", "-")
+                            .replace(")", "-"),
+                            "sample_name": "indexcheck",
+                            "sample_ref": "hg19",
+                        }
+
+                        added_dummy_samples.append(add_dummy_sample)
+
+            self.raw_samplesheet.extend(added_dummy_samples)
+
+            return self
+
+    def remove_unwanted_indexes(self):
+        """ Filter out indexes of unwanted length and single indexes """
+
+        if self.indexlength:
+            if self.pad and int(self.indexlength) in (16, 20):
+                self.raw_samplesheet = [
+                    line
+                    for line in self.raw_samplesheet
+                    if len(line["index"].replace("-", ""))
+                    in (16, int(self.indexlength))
+                ]
```

Comment:
I find these very hard to understand - there is simply to much going on. Try to avoid these onliners as they are hard to read and become quite complex. Use utility subs and small functional defs that you can test and label with what they are doing

Reply:
I really like list comprehensions, they are really pythonic and to me very few things beat `[expression for item in iterable if condition]` for readability. Moving the condition part to a function or method can help with readability though, especially in this case.

After consulting Anna Z I further simplified this method. No need for a lot of checks, we can only have dual8 or dual10 indexes, no single indexes. We pad everything that is not length 8 unless it's fluffy. SO just removing the single indexes leaves us with a correct samplesheet that includes the run samples and the dummy index samples of the correct length.

Reply:
Yes, but can you glance at this expression for 5 seconds and explain what it does? Can you modify it and be sure you did not alter the logic? I guess what i am trying to say is that if the comprehensions have very easy to understand `conditions`, then they are fine. But this one to me does not.

Reply:
this is what it looks like right now:
```
 def remove_unwanted_indexes(self) -> "CreateNovaseqSamplesheet":
        """ Filter out indexes of unwanted length and single indexes """

        self.raw_samplesheet = [
            line for line in self.raw_samplesheet if self.is_dual_index(line["index"])
        ]

        return self
```
And the condition is this:
```
@staticmethod
def is_dual_index(index: str, delimiter="-") -> bool:
    """ Determines if an index in the raw samplesheet is dual index or not """
    return delimiter in index
```

The weird looking indexlength conditions are not needed any more, we just need dual index only in the sheet. An index is a dual index if and only if it has a `-` in it.